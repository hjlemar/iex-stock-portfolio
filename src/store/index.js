import axios from 'axios';
import Vue from 'vue';
import Vuex from 'vuex';
import moment from 'moment';
import { CHANGE_DATE, STOCK_DATA_FOUND, QUERY_STOCK_DQUOTE,
  ADD_STOCK_RECENT_HISTORY, SET_CHART_DATA, QUERY_CHART } from './mutation-types';

Vue.use(Vuex);

export const storeConfig = {
  state: {
    date: moment(new Date()).format('YYYY-MM-DD'),
    data: null,
    stocks: [],
    chart: [],
  },
  mutations: {
    [CHANGE_DATE](state, payload) {
      state.date = payload.date;
    },
    [STOCK_DATA_FOUND](state, { payload }) {
      state.data = payload.data;
    },
    [ADD_STOCK_RECENT_HISTORY](state, { payload }) {
      const stock = payload.toUpperCase();
      const temp = state.stocks.filter(item => item !== stock);
      state.stocks = [stock, ...temp];
    },
    [SET_CHART_DATA](state, { payload }) {
      state.chart = [...payload];
    },
  },
  getters: {
    date(state) {
      return state.date;
    },
    data(state) {
      return state.data;
    },
    stocks(state) {
      return state.stocks;
    },
    chart(state) {
      return [...state.chart];
    },
  },
  actions: {
    [QUERY_STOCK_DQUOTE]({ dispatch, commit }, stock) {
      return axios.get(`https://api.iextrading.com/1.0/stock/${stock}/delayed-quote`)
        .then((payload) => {
          commit({ type: ADD_STOCK_RECENT_HISTORY, payload: stock });
          commit({ type: STOCK_DATA_FOUND, payload });
          return dispatch(QUERY_CHART, stock);
        });
    },
    [QUERY_CHART]({ commit }, stock) {
      return axios.get(`https://api.iextrading.com/1.0/stock/${stock}/chart`)
        .then(({ data }) => commit({ type: SET_CHART_DATA, payload: data }));
    },
  },
};

export default new Vuex.Store(storeConfig);
