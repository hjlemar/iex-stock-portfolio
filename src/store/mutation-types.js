export const CHANGE_DATE = 'changeDate';
export const STOCK_DATA_FOUND = 'stockDataFound';
export const QUERY_STOCK_DQUOTE = 'queryStockDelayedQuote';
export const ADD_STOCK_RECENT_HISTORY = 'addStockToRecentHistory';
export const SET_CHART_DATA = 'setChartData';
export const QUERY_CHART = 'queryChart';

