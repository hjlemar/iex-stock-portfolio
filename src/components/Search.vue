<template>
  <div>
    <v-layout row>
        <v-text-field
          name="stock-search"
          label="Stock Search"
          id="stock-search"
          v-model="stock"
          @keyup.enter="goToDetails()"
        ></v-text-field>
         <v-btn
          id='search-button'
          @click="goToDetails()"
          :disabled="!stock"
          color="primary"
          >Get Quote</v-btn>
    </v-layout>
    <v-layout row>
        <v-alert type="error" :value="failed" transition='slide-x-transition'>
          {{ message }}
        </v-alert>
    </v-layout>
  </div>
</template>

<script>
import { QUERY_STOCK_DQUOTE } from '../store/mutation-types';

export default {
  data() {
    return {
      stock: null,
      failed: false,
      message: null,
    };
  },
  methods: {
    goToDetails() {
      this.failed = false;

      this.$store.dispatch(QUERY_STOCK_DQUOTE, this.stock)
        .then(() => {
          this.stock = null;
          this.$router.push({ path: '/details' });
        })
        .catch((err) => {
          console.log(err);
          this.message = err.response.data;
          this.failed = true;
        });
    },
  },
};
</script>

<style>

</style>

