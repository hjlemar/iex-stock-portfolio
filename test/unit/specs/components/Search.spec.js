import { mount, shallow, createLocalVue } from '@vue/test-utils';
import { QUERY_STOCK_DQUOTE } from '@/store/mutation-types';
import Vuex from 'vuex';
import Search from '@/components/Search';
import Vuetify from 'vuetify';

const localVue = createLocalVue();
localVue.use(Vuex);
localVue.use(Vuetify);


describe('Search', () => {
  it('should render component and match snapshot', () => {
    const wrapper = shallow(Search, { store: {}, localVue });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should have a default state', () => {
    const wrapper = mount(Search, { store: {}, localVue });
    expect(wrapper.vm.stock).toBeNull();
    expect(wrapper.vm.failed).toBeFalsy();
    expect(wrapper.vm.message).toBeNull();
  });

  describe('goToDetails gets triggered by', () => {
    [{ event: 'click', id: 'search-button' },
      { event: 'keyup.enter', id: 'stock-search' },
    ].forEach(({ event, id }) => {
      it(`${event} event`, () => {
        const wrapper = mount(Search, { store: {}, localVue });
        const goToDetails = jest.fn();
        wrapper.setMethods({ goToDetails });
        const selector = `#${id}`;
        const elem = wrapper.find(selector);
        elem.trigger(event);
        expect(goToDetails).toBeCalled();
      });
    });
  });

  describe('gotToDetails', () => {
    let store;
    let actions;

    beforeEach(() => {
      actions = {
        [QUERY_STOCK_DQUOTE]: jest.fn(),
      };
      store = new Vuex.Store({
        actions,
      });
    });

    it('should dispatch QUERY_STOCK_DQUOTE event', () => {
      const push = jest.fn((x) => {
        expect(x).toMatchObject({ path: '/details' });
      });
      const mocks = {
        $router: {
          push,
        },
      };
      const wrapper = shallow(Search, {
        store,
        localVue,
        data: { stock: 'O' },
        mocks,
      });
      expect(wrapper.vm.stock).toBe('O');
      wrapper.vm.goToDetails();
      expect(actions[QUERY_STOCK_DQUOTE]).toBeCalled();
      // this fails oddly, even though the expect in the
      // push function is evaluated. (console.log shows this.)
      // expect(push).toBeCalled();
    });
  });
});
