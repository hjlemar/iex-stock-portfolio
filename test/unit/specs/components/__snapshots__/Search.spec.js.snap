// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render component and match snapshot 1`] = `
<div>
    <!---->
    <!---->
</div>
`;
